File: youtube-fetch/src/cdp-client.js
import { chromium } from "playwright";

/**
 * 通过 connectOverCDP 附着到已启动的 Chrome，在 youtube-transcript.ai 页面上下文中 fetch 文稿。
 */
export class CdpTranscriptClient {
  /**
   * @param {{ cdpUrl: string, siteUrl: string, timeoutMs: number, log: ReturnType<import('./logger.js').createLogger> }} opts
   */
  constructor(opts) {
    this.cdpUrl = opts.cdpUrl;
    this.siteUrl = opts.siteUrl;
    this.timeoutMs = opts.timeoutMs;
    this.log = opts.log;
    /** @type {import('playwright').Browser | null} */
    this.browser = null;
    /** @type {import('playwright').Page | null} */
    this.page = null;
    this.ready = false;
    /** @type {Promise<unknown>} */
    this.chain = Promise.resolve();
  }

  async connect() {
    this.log.info(`connectOverCDP → ${this.cdpUrl}`);
    this.browser = await chromium.connectOverCDP(this.cdpUrl, { timeout: 30_000 });
    const ctx = this.browser.contexts()[0];
    if (!ctx) throw new Error("CDP 浏览器无可用 context");

    const existing = ctx.pages().find((p) => {
      try {
        return p.url().includes("youtube-transcript.ai");
      } catch {
        return false;
      }
    });
    this.page = existing ?? (await ctx.newPage());

    if (!existing) {
      this.log.info(`打开 ${this.siteUrl}`);
      await this.page.goto(this.siteUrl, { waitUntil: "domcontentloaded", timeout: this.timeoutMs });
    } else {
      this.log.info(`复用已打开标签: ${this.page.url().slice(0, 120)}`);
    }

    this.ready = true;
    this.log.info("CDP 文稿客户端就绪");
  }

  /**
   * @param {string} videoId
   * @param {string | undefined} lang
   */
  async fetchTranscriptText(videoId, lang) {
    if (!this.page || !this.ready) throw new Error("CDP 尚未连接，请先调用 connect()");
    return this.#enqueue(() => this.#fetchOnce(videoId, lang));
  }

  /**
   * @param {() => Promise<T>} fn
   * @returns {Promise<T>}
   * @template T
   */
  #enqueue(fn) {
    const run = this.chain.then(fn);
    this.chain = run.catch(() => {});
    return run;
  }

  /**
   * @param {string} videoId
   * @param {string | undefined} lang
   */
  async #fetchOnce(videoId, lang) {
    const page = this.page;
    if (!page) throw new Error("无可用 page");

    const path = lang
      ? `/transcript/${videoId}.txt?lang=${encodeURIComponent(lang)}`
      : `/transcript/${videoId}.txt`;
    const url = `${this.siteUrl}${path}`;
    this.log.debug(`goto ${url}`);

    try {
      const resp = await page.goto(url, { waitUntil: "commit", timeout: this.timeoutMs });
      if (!resp) throw new Error("page.goto 无响应");
      const status = resp.status();
      const text = await resp.text();
      if (status < 200 || status >= 300) {
        const hint = text ? ` — ${text.slice(0, 300)}` : "";
        throw new Error(`youtube-transcript.ai HTTP ${status}${hint}`);
      }
      if (!text.trim()) throw new Error("返回正文为空");
      return text;
    } catch (e) {
      const err = /** @type {Error} */ (e);
      if (err.message.includes("Target closed") || err.message.includes("has been closed")) {
        this.ready = false;
        throw new Error(`CDP 页面已关闭: ${err.message}`);
      }
      throw err;
    }
  }

  async close() {
    this.ready = false;
    this.page = null;
    if (this.browser) {
      await this.browser.close().catch(() => {});
      this.browser = null;
      this.log.info("已断开 CDP（未关闭你的 Chrome）");
    }
  }
}

File: youtube-fetch-py/youtube_fetch/cdp_client.py
"""通过 Playwright connect_over_cdp 在 youtube-transcript.ai 页面拉取文稿."""

from __future__ import annotations

import asyncio
import logging

from playwright.async_api import Browser, Page, async_playwright


class CdpTranscriptClient:
    def __init__(self, cdp_url: str, site_url: str, timeout_ms: int, log: logging.Logger) -> None:
        self.cdp_url = cdp_url
        self.site_url = site_url.rstrip("/")
        self.timeout_ms = timeout_ms
        self.log = log
        self._playwright = None
        self.browser: Browser | None = None
        self.page: Page | None = None
        self.ready = False
        self._serial = asyncio.Lock()

    async def connect(self) -> None:
        self.log.info("connect_over_cdp → %s", self.cdp_url)
        self._playwright = await async_playwright().start()
        self.browser = await self._playwright.chromium.connect_over_cdp(
            self.cdp_url, timeout=30_000
        )
        contexts = self.browser.contexts
        if not contexts:
            raise RuntimeError("CDP 浏览器无可用 context")
        ctx = contexts[0]

        existing = None
        for p in ctx.pages:
            try:
                if "youtube-transcript.ai" in p.url:
                    existing = p
                    break
            except Exception:
                continue

        self.page = existing or await ctx.new_page()
        if not existing:
            self.log.info("打开 %s", self.site_url)
            await self.page.goto(
                self.site_url,
                wait_until="domcontentloaded",
                timeout=self.timeout_ms,
            )
        else:
            self.log.info("复用已打开标签: %s", self.page.url[:120])

        self.ready = True
        self.log.info("CDP 文稿客户端就绪")

    async def fetch_transcript_text(self, video_id: str, lang: str | None = None) -> str:
        if not self.page or not self.ready:
            raise RuntimeError("CDP 尚未连接，请先调用 connect()")
        async with self._serial:
            return await self._fetch_once(video_id, lang)

    async def _fetch_once(self, video_id: str, lang: str | None) -> str:
        page = self.page
        if not page:
            raise RuntimeError("无可用 page")

        path = (
            f"/transcript/{video_id}.txt?lang={lang}"
            if lang
            else f"/transcript/{video_id}.txt"
        )
        url = f"{self.site_url}{path}"
        self.log.debug("goto %s", url)

        try:
            resp = await page.goto(url, wait_until="commit", timeout=self.timeout_ms)
            if not resp:
                raise RuntimeError("page.goto 无响应")
            status = resp.status
            text = await resp.text()
            if status < 200 or status >= 300:
                hint = f" — {text[:300]}" if text else ""
                raise RuntimeError(f"youtube-transcript.ai HTTP {status}{hint}")
            if not text.strip():
                raise RuntimeError("返回正文为空")
            return text
        except Exception as e:
            msg = str(e)
            if "Target closed" in msg or "has been closed" in msg:
                self.ready = False
                raise RuntimeError(f"CDP 页面已关闭: {msg}") from e
            raise

    async def close(self) -> None:
        self.ready = False
        self.page = None
        if self.browser:
            try:
                await self.browser.close()
            except Exception:
                pass
            self.browser = None
        if self._playwright:
            try:
                await self._playwright.stop()
            except Exception:
                pass
            self._playwright = None
        self.log.info("已断开 CDP（未关闭你的 Chrome）")
